mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala

```diff
@@ -46,19 +47,59 @@ class VectorAssembler(override val uid: String)
   def setOutputCol(value: String): this.type = set(outputCol, value)
 
   override def transform(dataset: DataFrame): DataFrame = {
+    // Schema transformation.
+    val schema = dataset.schema
+    lazy val first = dataset.first()
+    val attrs = $(inputCols).flatMap { c =>
```

Comment:
The indices gets reset in `AttributeGroup`: 

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/attribute/AttributeGroup.scala#L33

Let's re-visit the attribute API in 1.5:)